### PR TITLE
Render graph cube map

### PR DESCRIPTION
**Pull request by norda97**
Closes: 
- Cube Map
  - Add texture cube array support
  - Add texturecube support in rendergraph
- Rendergraph Polish
  - Add shader resource recursive folder loading
  - Add Push Constants to RenderGraph

Also fixes:
- Window resizing
- Rendergraph error reporting

### CODE REVIEW

File: LambdaEngine/Include/Rendering/Core/API/SBT.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+#include "GraphicsTypes.h"
+
+namespace LambdaEngine
+{
+	class SBT : public DeviceChild
+	{
+	public:
+		DECL_DEVICE_INTERFACE(SBT);
+
+		virtual void
```
Comment (jonathan2222) [inline LambdaEngine/Include/Rendering/Core/API/SBT.h]: This seems weird.  virtual void should be removed.

**Review by jonathan2222 — APPROVED** [review]
Seems good.

File: Sandbox/Source/Sandbox.cpp [hunk 2]
```diff
@@ -436,26 +481,48 @@ void Sandbox::Render(LambdaEngine::Timestamp delta)
...
 	}
 
+	//float32 test = Random::Float32();
+
+	//PushConstantsUpdate pushConstantsUpdateTest = {};
+	//pushConstantsUpdateTest.RenderStageName		= "DEMO";
+	//pushConstantsUpdateTest.pData				= &test;
+	//pushConstantsUpdateTest.DataSize			= sizeof(float32);
+	//Renderer::GetRenderGraph()->UpdatePushConstants(&pushConstantsUpdateTest);
+
```
Comment (TheoBerlin) [inline Sandbox/Source/Sandbox.cpp:491]:
```suggestion
```

Remove commented code

File: LambdaEngine/Source/Resources/ResourceManager.cpp [hunk 3]
```diff
@@ -259,6 +259,48 @@ namespace LambdaEngine
...
+		if (loadedTextureGUID != s_TextureNamesToGUIDs.end())
+			return loadedTextureGUID->second;
+
+		uint32 textureCount = count * 6U;
+
+		GUID_Lambda guid = GUID_NONE;
+		Texture** ppMappedTexture = nullptr;
+		TextureView** ppMappedTextureView = nullptr;
+
+		//Spinlock
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Resources/ResourceManager.cpp:274]: What is this comment about? I don't see a spinlock being used.

File: LambdaEngine/Source/Resources/ResourceLoader.cpp [hunk 4]
```diff
@@ -618,6 +618,57 @@ namespace LambdaEngine
...
 	}
 
+	Texture* ResourceLoader::LoadCubeTexturesArrayFromFile(const String& name, const String& dir, const String* pFilenames, uint32 count, EFormat format, bool generateMips)
+	{
+		int texWidth = 0;
+		int texHeight = 0;
+		int bpp = 0;
+
+		const uint32 textureCount = count;
+		TArray<void*> stbi_pixels(textureCount);
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Resources/ResourceLoader.cpp:628]: Isn't this an array of textures? The naming suggests it's an array of pixels.

File: LambdaEngine/Source/Resources/ResourceLoader.cpp [hunk 5]
```diff
@@ -618,6 +618,57 @@ namespace LambdaEngine
...
+		if (format == EFormat::FORMAT_R8G8B8A8_UNORM)
+		{
+			uint32 flags = FTextureFlags::TEXTURE_FLAG_CUBE_COMPATIBLE | FTextureFlags::TEXTURE_FLAG_SHADER_RESOURCE;
+			pTexture = LoadTextureArrayFromMemory(name, stbi_pixels.GetData(), stbi_pixels.GetSize(), texWidth, texHeight, format, flags, generateMips);
+		}
+
+		for (uint32 i = 0; i < textureCount; i++)
+		{
+			stbi_image_free(stbi_pixels[i]);
+		}
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Resources/ResourceLoader.cpp:667]:
```suggestion
		for (void* pTexture : stbi_pixels)
		{
			stbi_image_free(pTexture);
		}
```

If you choose to rename the array (stbi_pixels), don't forget to rename it here as well.

File: LambdaEngine/Source/Resources/ResourceLoader.cpp [hunk 6]
```diff
@@ -618,6 +618,57 @@ namespace LambdaEngine
 		return pTexture;
 	}
 
+	Texture* ResourceLoader::LoadCubeTexturesArrayFromFile(const String& name, const String& dir, const String* pFilenames, uint32 count, EFormat format, bool generateMips)
+	{
+		int texWidth = 0;
+		int texHeight = 0;
+		int bpp = 0;
+
+		const uint32 textureCount = count;
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Resources/ResourceLoader.cpp:627]: Can't the count variable be used? Why should it be copied?

File: LambdaEngine/Source/Rendering/RenderGraphParser.cpp [hunk 7]
```diff
@@ -120,6 +120,7 @@ namespace LambdaEngine
 		}
 
 		//Loop Through each Render Stage in Order and create synchronization stages
+		TSet<String> resourceNamesActuallyUsed;
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Rendering/RenderGraphParser.cpp:123]: I only see strings being inserted into the set; it's never read from. Remove? Also, use an `unordered_set` unless you want the elements sorted.

File: LambdaEngine/Source/Rendering/Core/Vulkan/TextureViewVK.cpp [hunk 8]
```diff
@@ -92,6 +97,15 @@ namespace LambdaEngine
 		}
 	}
 
+	void TextureViewVK::InternalRelease()
+	{
+		if (m_ImageView != VK_NULL_HANDLE)
+		{
+			SAFERELEASE(m_Desc.pTexture);
```
Comment (TheoBerlin) [inline LambdaEngine/Source/Rendering/Core/Vulkan/TextureViewVK.cpp:104]: Does `SAFERELEASE` need to be within the if-statement? If it's `nullptr`, nothing will happen anyways.

File: LambdaEngine/Include/Rendering/RenderGraphTypes.h [hunk 9]
```diff
@@ -654,6 +667,32 @@ namespace LambdaEngine
...
+
+		if (pRenderStageDesc->Type == EPipelineStateType::PIPELINE_STATE_TYPE_GRAPHICS)
+		{
+			if (pRenderStageDesc->Graphics.Shaders.PixelShaderName.size() > 0)		pipelineStageMask |= FPipelineStageFlags::PIPELINE_STAGE_FLAG_PIXEL_SHADER;
+			if (pRenderStageDesc->Graphics.Shaders.DomainShaderName.size() > 0)		pipelineStageMask |= FPipelineStageFlags::PIPELINE_STAGE_FLAG_DOMAIN_SHADER;
+			if (pRenderStageDesc->Graphics.Shaders.HullShaderName.size() > 0)		pipelineStageMask |= FPipelineStageFlags::PIPELINE_STAGE_FLAG_HULL_SHADER;
+			if (pRenderStageDesc->Graphics.Shaders.GeometryShaderName.size() > 0)	pipelineStageMask |= FPipelineStageFlags::PIPELINE_STAGE_FLAG_GEOMETRY_SHADER;
+			if (pRenderStageDesc->Graphics.Shaders.VertexShaderName.size() > 0)		pipelineStageMask |= FPipelineStageFlags::PIPELINE_STAGE_FLAG_VERTEX_INPUT;
+			if (pRenderStageDesc->Graphics.Shaders.MeshShaderName.size() > 0)		pipelineStageMask |= FPipelineStageFlags::PIPELINE_STAGE_FLAG_MESH_SHADER;
+			if (pRenderStageDesc->Graphics.Shaders.TaskShaderName.size() > 0)		pipelineStageMask |= FPipelineStageFlags::PIPELINE_STAGE_FLAG_TASK_SHADER;
```
Comment (TheoBerlin) [inline LambdaEngine/Include/Rendering/RenderGraphTypes.h:682]: Consider checking `std::string::empty()` instead of `std::string::size()`.

**Review by TheoBerlin — CHANGES_REQUESTED** [review]

**Pull request merged**